### ISSUE TITLE
headless mode: change selection

### ISSUE THREAD

**Issue by rdaems** [annotorious/annotorious#80]
Hi, 

first of all thanks for your great work! I was looking for a simple, lightweight annotator that I could easily include in my existing tools.
This looks perfect.

I'm trying to figure out how `headless` mode works, it doesn't seem to be documented yet?

So I did this:

```
var anno = Annotorious.init({
    image: 'hallstatt',
    locale: 'auto',
    headless: true
})
```

And figured out I had to add code to store new labels:

```
anno.on('createSelection', function(selection) {
  selection.body = [ body ];
  anno.updateSelected(selection, true);
});
```

What I have trouble with now, is to preserve changes to existing annotations. If I change a label, the changes are not persisten. The label reverts to the state before I started making changes. I tried this but it doesn't work:

```
anno.on('changeSelectionTarget', function(target) {
  anno.updateSelected(target, true);
});
```

In summary, I just need a tool for quickly drawing polygons, without defining classes (all labels are the same class). And would love the option of making changes afterwards.

Thanks for your time!

**Comment by rsimon** [comment]
Thanks the kind feedback! To be honest: headless mode is still in a state of development :-) How it's __supposed__ to work is, pretty much, like you describe.

- You set `headless: true` on init - this will prevent the popup from opening when the user creates a selection
- You use a handler on `createSelection` to inject your own body into the annotation - just like you did

None of this should open the editor. Annotorious should just fire a `createAnnotation` event with the newly created annotation (and the body you have injected) right after you called `.updateSelected`. Was this working so far? 

The `changeSelectionTarget` shouldn't be relevant. It fires when the popup is open and the user moves or resizes the shape. There's also a bit of an inconsistency with the other events insofar as you will get a `target` object (representing the annotated region), whereas the other events generally return the whole `annotation` object. (Or the `selection`, which has basically the same form as the annotation, just with a different `type` to show that it's still in draft state.)

If things __do not__ work this way, it's very likely a bug in Annotorious. I can take a look this weekend. Do you perhaps have full code snippet of what you have so far, so I can retrace your steps?

**Comment by rdaems** [comment]
Hi Rainer,

Yes creating a new annotation and saving it explicitly with the `createSelection` handler works as expected! (though I didn't connect a backend yet, but I wouldn't expect issues relating to this specific case).

The problem is then, **changing an existing annotation** (moving it or moving individual points of the polygon). This does not work, the changes are not persistent.

This is my code:

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>Annotorious | Basic Example</title>
    <link href="https://fonts.googleapis.com/css?family=Lato&display=swap" rel="stylesheet">
    <style>
      html, body {
        padding:20px;
        margin:0px;
        font-family:'Lato', sans-serif;
      }
      
      #content {
        width:100%;
      }

      h1 {
        font-size:21px;
        font-weight:normal;
        margin:0;
        padding:0;
      }

      p.instructions {
        padding:10px 0;
      }

      img {
        max-width:100%;
      }

      p.caption {
        font-family:Arial, Helvetica, sans-serif;
        color:#8f8f8f;
      }

      p.caption a {
        color:#3f3f3f;
      }
    </style>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@recogito/annotorious@latest/dist/annotorious.min.css">
    <script src="https://cdn.jsdelivr.net/npm/@recogito/annotorious@latest/dist/annotorious.min.js"></script>
  </head>
  <body>
    <div id="content">    
      <h1>Annotorious: Basic Example</h1>  
      <p class="instructions">
        Click the annotation to edit. Click and drag the mouse to create a new annotation.
      </p>
      <p>
        <button id="current-tool">RECTANGLE</button>
      </p>
      <img id="hallstatt" src="640px-Hallstatt.jpg">
      <p class="caption">
        Sample image source: <a href="http://commons.wikimedia.org/wiki/File:Hallstatt_300.jpg">Hallstatt,
        Austria</a>, by Nick Csakany/Wikimedia Commons. Public Domain.
      </p>
    </div>
  </div>
    <script>
      (function() {
        var anno = Annotorious.init({
          image: 'hallstatt',
          locale: 'auto',
          headless: true
        });

        var body = {
          type: 'TextualBody',
          purpose: 'tagging',
          value: 'wheal'
        };

        anno.on('createSelection', function(selection) {
          console.log('createSelection');
          selection.body = [ body ];
          console.log(selection)
          anno.updateSelected(selection, true);
        });

        // anno.deleteSelected() 
        // anno.updateSelected(body)
        // anno.updateSelected() // updates only the target

        anno.on('changeSelectionTarget', function(target) {
          console.log('selection changed', target);
          var selected = anno.getSelected();
          console.log(selected)
          anno.updateSelected(selected, true);
        });

        anno.on('selectAnnotation', function(annotation) {
          console.log('selected', annotation);
        });

        anno.on('createAnnotation', function(a) {
          console.log('created', a);
        });

        anno.on('updateAnnotation', function(annotation, previous) {
          console.log('updated', previous, 'with', annotation);
          anno.updateSelected(annotation, true);
        });

        anno.on('deleteAnnotation', function(annotation) {
          console.log('deleted', annotation);
        });
        
        // anno.on('mouseEnterAnnotation', function(annotation) {
        //   console.log('enter', annotation);
        // });
        
        // anno.on('mouseLeaveAnnotation', function(annotation) {
        //   console.log('leave', annotation);
        // });

        anno.loadAnnotations('annotations.w3c.json');

        anno.setDrawingTool('polygon');
      })()
    </script>
  </body>
</html>
```

**Comment by rsimon** [comment]
Ah, I think I understand now what's causing the confusion & problems. Yes, it is as I said: mostly a consequence of Annotorious being unfinished in this regard. The headless mode pre-dates editable shapes. I.e. it's easy to make things work - until users select a shape and edit it, as you noticed.

The root cause is that I have not yet come up with a good way to handle the lifecycle for headless editing. In "non-headless" mode, the user selects the annotation. Editor pops up, the user can move and resize, clicks Ok -> changes get persisted and the `updateAnnotation` event fires.

In headless mode, there's no effective "Ok" activity. I assume that the `updateAnnotation` event should either fire continuously as the user drags and resizes the annotation; or it should fire as soon as the user (implicitely) indicates they are done editing by deselecting the shape. (The latter option is probably the better one.)

Right now this should _almost_ work, I think. But probably the deselection causes a "cancel" behaviour (erasing all changes) rather than an "ok" behavior.

Anyways - as I said I'll try to find some time over the weekend to come up with a solution. After that, the goal would be to make the following code code work:

```js
var anno = Annotorious.init({
  image: 'hallstatt',
  locale: 'auto',
  headless: true
});

var body = {
  type: 'TextualBody',
  purpose: 'tagging',
  value: 'wheal'
};

// User draws a *new* shape - inject my tag
anno.on('createSelection', function(selection) {
  selection.body = [ body ];
  anno.updateSelected(selection, true);
});

anno.on('createAnnotation', function(annotation) {
  // The user has finished drawing a new shape, tag was injected
  // TODO: create the new annotation on the server
});

anno.on('updateAnnotation', function(annotation, previous) {
  // The user has resized or moved a shape
  // TODO update the previous annotation on the server with the new one
});

anno.on('selectAnnotation', function(annotation) {
  // Just to keep track of user selections. For example, there could be an external "DELETE" button 
  // somewhere in my app to programmatically remove the annotation from Annotorious and delete it
  // from the server, if the user clicks that button now.
});
```

I hope my attempt to explain makes halfway sense :-)

**Comment by rsimon** [comment]
Hi, I tied up a few lose ends in the code now & things now work as described above:

- Select a shape by clicking it
- Move/resize the shape
- As soon as you deselect the shape (either by clicking outside, or by selecting another shape), the `updateAnnotation` event is fired properly
- In addition: after you have moved/resized, pressing `Escape` will function as cancel. The shape will pop back into its original state, no event will be fired

You can use this now if you build the code from source. But I'll also issue a new release maybe later today or tomorrow.

**Comment by rsimon** [comment]
New release is up & published to npm: https://github.com/recogito/annotorious/releases/tag/v2.2.1

**Comment by rdaems** [comment]
Hi Rainer,

Great! Thanks a lot for your quick response!
It seems to work fine now, so I'll close the issue :)

All the best
